### PR TITLE
add ligand, water stripping and save the selected atoms as CIF in synthetic density generation

### PR DESCRIPTION
**Pull request by k-chrispens**
## Summary by CodeRabbit

* **New Features**
  * CLI flags to remove hydrogens (excluding deuterium), waters, and ligands before density generation, plus optional saving of preprocessed structures; coordinate alignment applied when saving.
  * Utilities to filter atom sets to polymer or amino-acid atoms.

* **Improvements**
  * Occupancy handling validates inputs, forces custom mode when custom values supplied, and emits warnings on mismatches.

* **Bug Fixes**
  * Safer save operations with error handling and logged failures.

* **Tests**
  * Expanded tests for new utilities, filters, edge cases, and representation parity.

### CODE REVIEW

File: src/sampleworks/eval/generate_synthetic_density.py [hunk 1]
```diff
@@ -235,6 +247,16 @@ def _process_single_row(
...
     device
         PyTorch device for computation
+    strip_hydrogens
+        If True, remove hydrogen atoms before computing density.
+    strip_waters
+        If True, remove water molecules before computing density.
+    strip_ligands
+        If True, remove ligand molecules (non-water heteroatoms) before computing density. This is
+        done by keeping only polymer atoms in the selection, which are typically not ligands.
+        TODO: be more thorough with this? We could make this a transform
```
Comment (marcuscollins) [inline src/sampleworks/eval/generate_synthetic_density.py]: Yes.

File: src/sampleworks/eval/generate_synthetic_density.py [hunk 2]
```diff
@@ -235,6 +247,16 @@ def _process_single_row(
         If True, use electron scattering factors. If False, use X-ray factors.
     device
         PyTorch device for computation
+    strip_hydrogens
+        If True, remove hydrogen atoms before computing density.
```
Comment (marcuscollins) [inline src/sampleworks/eval/generate_synthetic_density.py]: good practice to note defaults.

File: src/sampleworks/eval/generate_synthetic_density.py [hunk 3]
```diff
@@ -256,13 +278,37 @@ def _process_single_row(
         logger.error(f"Selection error for {row.filename}: {e}")
         return
 
-    altloc_info = detect_altlocs(atom_array)
+    if strip_hydrogens:
```
Comment (marcuscollins) [inline src/sampleworks/eval/generate_synthetic_density.py]: minor style nit: I would one-line simple ifs like this: `atom_array = remove_hydrogens(atom_array) if strip_hydrogens else atom_array`

File: src/sampleworks/eval/generate_synthetic_density.py [hunk 4]
```diff
@@ -213,18 +218,25 @@ def load_batch_csv(csv_path: Path) -> list[BatchRow]:
 
 def _process_single_row(
     row: BatchRow,
```
[the remark this answers is not in the document]
Comment (marcuscollins) [reply src/sampleworks/eval/generate_synthetic_density.py]: I think I can easily enough implement the needed changes to this code as well to handle multiple selections if that's desirable.

[the remark this answers is not in the document]
Comment (marcuscollins) [reply src/sampleworks/eval/generate_synthetic_density.py]: https://github.com/diff-use/sampleworks/issues/97

File: src/sampleworks/eval/generate_synthetic_density.py [hunk 5]
```diff
@@ -396,19 +502,33 @@ def main() -> None:
...
+            strip_waters=args.remove_waters,
+            strip_ligands=args.remove_ligands,
+            save_structure=args.save_structure,
         )
     elif args.structure:
         atom_array = load_structure_with_altlocs(args.structure)
         atom_array = apply_selection(atom_array, args.selection)
 
-        altloc_info = detect_altlocs(atom_array)
+        if args.remove_hydrogens:
```
Comment (marcuscollins) [inline src/sampleworks/eval/generate_synthetic_density.py]: Could you call `_process_single_row` here instead?

**Review by marcuscollins — CHANGES_REQUESTED** [review]
Mostly a few comments for discussion and some nits. I would suggest re-using the method `_process_single_row` instead of duplicating its code in `main`. It is a complicated enough method that I think the benefit of keeping it in one place is worth it.

File: src/sampleworks/eval/generate_synthetic_density.py (hunk 1 above)
Comment (k-chrispens) [reply src/sampleworks/eval/generate_synthetic_density.py]: #99

File: tests/utils/test_atom_array_utils.py [hunk 6]
```diff
@@ -305,3 +313,137 @@ def test_preserves_coordinates(self, atom_array_partial_overlap):
...
+        once = remove_hydrogens(_get_atom_array(test_structure))
+        twice = remove_hydrogens(once)
+        assert len(once) == len(twice)
+
+    def test_invalid_type_raises_error(self):
+        with pytest.raises(TypeError, match="can only accept AtomArray or AtomArrayStack"):
+            remove_hydrogens("bad input")  # pyright: ignore[reportArgumentType]
+
+
+class TestKeepPolymer:
```
[the remark this answers is not in the document]
Comment (k-chrispens) [reply tests/utils/test_atom_array_utils.py]: The usage of both keep_polymer, etc. is kind of a hack, because what I really need is a way of keeping all the protein chain including non-canonical residues. Currently the functions applied in series accomplish this, and I couldn't figure out a way to do it another way. There is probably a better way, though. Will add comment

File: tests/utils/test_atom_array_utils.py [hunk 7]
```diff
@@ -305,3 +313,137 @@ def test_preserves_coordinates(self, atom_array_partial_overlap):
...
+        """AtomArray and AtomArrayStack paths must select the same atoms."""
+        arr_result = remove_hydrogens(_get_atom_array(test_structure))
+        stack_result = remove_hydrogens(_get_atom_array_stack(test_structure))
+
+        assert isinstance(stack_result, AtomArrayStack)
+        stack_result = cast(AtomArrayStack, stack_result)
+        assert len(arr_result) == stack_result.array_length()
+        assert "H" not in cast(np.ndarray, stack_result.element)
+
+    def test_is_idempotent(self, test_structure):
```
[the remark this answers is not in the document]
Comment (k-chrispens) [reply tests/utils/test_atom_array_utils.py]: Would you just rename the test?

File: src/sampleworks/eval/generate_synthetic_density.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (marcuscollins) [reply src/sampleworks/eval/generate_synthetic_density.py]: Once this is merged, I'll get my multi-selection PR in and update this as well.

**Review by marcuscollins — APPROVED** [review]
Thanks for the changes. LGTM
